File: pending/20.py
# 20. Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        open_ch = ('(', '[', '{')
        pair = {
            '(': ')', 
            '{': '}',
            '[': ']'
            }
        for i in range(len(s)):
            if s[i] in open_ch:
                stack.append(s[i])
            else:
                if stack and pair[stack[-1]] == s[i]:
                    stack.pop()
                else:
                    return False
        if stack:
            return False
        return True
        

File: algorithms/swift/20.swift
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/description/

class Solution {
    func isValid(_ s: String) -> Bool {
        let openBrackets: Set<Character> = ["(", "[", "{"]
        let data: [Character:Character] = [
            ")": "(", 
            "]": "[",
            "}": "{"
        ]

        var stack: [Character] = []

        for ch in s {
            if openBrackets.contains(ch) {
                stack.append(ch)
            }
            else if let need = data[ch] {
                if stack.last != need { return false }
                stack.removeLast()
            }
            else {
                return false
            }
        }
        return stack.isEmpty
    }
}
